muspy/processors/event.py

```diff
@@ -0,0 +1,394 @@
+VELOCITY = "velocity"      # velocity(track_id, velocity)
+PROGRAM = "program"        # program(track_id, program)
+DRUM = "drum"              # drum(track_id)
+TIME_SHIFT = "time_shift"  # time_shift(ticks)
+EOS = "eos"                # eos()
+
+ALL_NOTES = -1
+
+
+class EventRepresentationProcessor:
+    """Event-based representation processor.
+    The event-based represetantion represents music as a sequence of
+    events, including note-on, note-off, time-shift and velocity events.
+    The output shape is M x 1, where M is the number of events. The
+    values encode the events. The default configuration uses 0-127 to
+    encode note-one events, 128-255 for note-off events, 256-355 for
+    time-shift events, and 356 to 387 for velocity events.
+    Attributes
+    ----------
+    use_single_note_off_event : bool
+        Whether to use a single note-off event for all the pitches. If
+        True, the note-off event will close all active notes, which can
+        lead to lossy conversion for polyphonic music. Defaults to
+        False.
+    use_end_of_sequence_event : bool
+        Whether to append an end-of-sequence event to the encoded
+        sequence. Defaults to False.
+    encode_velocity : bool
+        Whether to encode velocities.
+    force_velocity_event : bool
+        Whether to add a velocity event before every note-on event. If
+        False, velocity events are only used when the note velocity is
+        changed (i.e., different from the previous one). Defaults to
+        True.
+    max_time_shift : int
+        Maximum time shift (in ticks) to be encoded as an separate
+        event. Time shifts larger than `max_time_shift` will be
+        decomposed into two or more time-shift events. Defaults to 100.
+    velocity_bins : int
+        Number of velocity bins to use. Defaults to 32.
+    default_velocity : int
+        Default velocity value to use when decoding. Defaults to 64.
+    encode_instrument: bool
+        Whether to encode the `program` and `is_drum` attributes for
+        each track. Defaults to False.
+    default_program: int
+        Default `program` value to use when decoding. Defaults to 0.
+    default_is_drum: bool
+        Default `is_drum` value to use when decoding. Defaults to
+        False.
+    num_tracks: int or None
+        The maximum number of tracks. Defaults to None, which means
+        single-track mode (encode all events as if they were in one
+        track).
+    ignore_empty_tracks: bool
+        Whether empty tracks should be ignored when encoding and deleted
+        when decoding. Defaults to False.
+    resolution: int
+        Time steps per quarter note to use when decoding. Defaults to
+        `muspy.DEFAULT_RESOLUTION`.
+    force_resolution: bool
+        Whether to adjust the resolution of the music when encoding.
+        If False and if the resolution is not correct, an error will be
+        raised. Defaults to False.
```

Comment:
This should also be included in other representation processors. 👍

Reply:
The question is whether to copy. Copying seems to be slow, but modifying the input maybe unexpected. So I would go with: warn by default (with `force_resolution=False`) and otherwise modify in place, but document this.

Reply:
I see. That's an issue. Calling `adjust_resolution` to modify the data in place is not so obvious from the argument name (`force_adjust_resolution` is better if we want this), but it still doesn't feel right to modify the data. I probably won't go for copying, so maybe we shouldn't support this? We could instead have a `check_resolution` (True by default) argument that assert the resolution is correct.